File: exercises/homework_command_line_program/js/cat.js
const fs = require('fs');
const readline = require('readline');

/**
 * Opens a file and scans its lines, handling flags for formatted output.
 * @param {string} fileName - The name of the file to open and scan.
 * @param {string} flag - The flag to determine output format ('', '-n', '-nb').
 */
const openAndScanFile = async (fileName, flag) => {
    try {
        const fileStream = fs.createReadStream(fileName);

        const rl = readline.createInterface({
            input: fileStream,
            crlfDelay: Infinity
        });

        let lineNumber = 1;

        for await (const line of rl) {
            switch (flag) {
                case '':
                    console.log(line);
                    break;
                case '-n':
                    console.log(`${lineNumber}. ${line}`);
                    lineNumber++;
                    break;
                case '-nb':
                    if (line.trim() === '') {
                        console.log('');
                    } else {
                        console.log(`${lineNumber}. ${line}`);
                        lineNumber++;
                    }
                    break;
                default:
                    console.log("Flag not recognized, please use --help to see documentation :D");
                    rl.close();
                    return;
            }
        }
    } catch (error) {
        console.error(`Unable to read file: ${error.message}`);
    }
}


 const cat = async (args) => {
    if (args.length > 0) {
        const fileNameFromCommandLine = args[0];
        if (args.length === 1) {
            await openAndScanFile(fileNameFromCommandLine, '');
        } else if (args.length === 2) {
            await openAndScanFile(fileNameFromCommandLine, args[1]);
        }
    } else {
        console.log("Please provide a file name.");
    }
}

module.exports = cat

File: exercises/homework_command_line_program/cat.go
package homework_command_line_program

import (
	"bufio"
	"fmt"
	"io"
	"log"
	"os"
)

// Options to read file
// 1. os.ReadFile
// 2. os.Open
// 3. use 'bufio' to scan text line by line

func OpenAndReadFile(fileName string) {
	file, err := os.Open(fileName)

	if err != nil {
		log.Fatalf("Unable to read file: %v", err)
	}
	defer file.Close()

	buffer := make([]byte, 1024)

	for {
		bytesRead, err := file.Read(buffer)

		if err == io.EOF {
			break
		}

		if err != nil {
			fmt.Println(err)
			continue
		}

		text := string(buffer[:bytesRead])

		if bytesRead > 0 {
			fmt.Printf("%s", text)
		}
	}
}

func OpenAndScanFile(fileName string, flag string) {
	file, err := os.Open(fileName)

	if err != nil {
		log.Fatalf("Unable to read file: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	lineNumber := 1

	for scanner.Scan() {
		line := scanner.Text()

		switch flag {
		case "":
			fmt.Printf("%s", line)
		case "-n":
			fmt.Printf("%d. %s\n", lineNumber, line)
			lineNumber++
		case "-nb":
			if line == "" {
				fmt.Printf("%s\n", line)
			} else {
				fmt.Printf("%d. %s\n", lineNumber, line)
				lineNumber++
			}
		default:
			fmt.Println("Flag not recognized, please use --help to see documentation :D ")
			return
		}

	}
}

func Cat(args []string) {
	//args := os.Args[1:]
	if len(args) > 0 {
		//fileNameFromCommandLine := strings.Join(args, " ")
		fileNameFromCommandLine := args[0]
		//OpenAndReadFile(fileNameFromCommandLine)
		if len(args) == 1 {
			OpenAndScanFile(fileNameFromCommandLine, "")
		} else if len(args) == 2 {
			OpenAndScanFile(fileNameFromCommandLine, args[1])
		}
	}

}
